### ISSUE TITLE
dump.rdb getting set to read-only

### ISSUE THREAD

**Issue by jeffutter** [docker-library/redis#44]
I seem to be having a recurring issue. Every few days my redis container starts throwing errors such as:

```
1:M 04 Feb 17:45:03.166 # Background saving error
1:M 04 Feb 17:45:09.085 * 1 changes in 3600 seconds. Saving...
1:M 04 Feb 17:45:09.085 * Background saving started by pid 1031
1031:C 04 Feb 17:45:09.086 # Failed opening .rdb for saving: Permission denied
```

Which is causing my clients to fail writes.

I have docker exec'd into the container and verified that the /data/dump.rdb file is read-only. I imagine this is causing those errors. I can't figure out why that file would be read-only.

I am running the container with no options other than container names and ports. I'm not mounting in any volumes or anything of the sort.

Anyone have any ideas?

**Comment by bsodmike** [comment]
+1 @jeffutter this has also become an issue on my box, and having `exec`'d inside the container I've confirmed the user is still `redis`.  On my box, I've linked `/data` to the host and that's showing a u:g of `999:docker` &mdash; and that doesn't matter as whilst exec'd into the container if I su into the redis user I'm able to create files, update, and even delete them.

Not sure what's causing this.  For the time being deleting the `dump.rdb` and restarting the container keeps it running for a shortwhile until it has to persist stuff to disk again.

**Comment by tianon** [comment]
What happens if you `chmod +w` the `dump.rdb` file after this occurs?  Does that allow Redis to start back up?  It sounds like Redis itself is creating the file without write permission, which is strange. :confused:

**Comment by bsodmike** [comment]
@tianon thought I tried that a while back, but I'll make sure to try that if it happens again.  So far it has been running without issues.

**Comment by timanovsky** [comment]
Same here. It is not file permissions, it is directory permissions.
```
docker run -it -v /Users/timanovsky/src/workato/kubernetes/docker/redis/data:/data  -u redis redis:3-32bit bash
redis@8517c376f55b:/data$ whoami
redis
redis@8517c376f55b:/data$ echo 123 > a
bash: a: Permission denied
```

**Comment by yosifkit** [comment]
@timanovsky, that looks like a different issue and might be related to https://github.com/boot2docker/boot2docker/issues/581.  In order to use a shared folder from OSX/Windows, you have to run the container with the correct user.  Many of the official images can be run with `--user 1000:50` to be able to access files shared from OSX/Windows.

```console
$ docker run -d -v /home/redis/data:/data -u 1000:50 redis:3-32bit
```

If you are running directly on Linux, then the entrypoint script will chown the bind mounted directory on start and then step down from `root` to `redis`: https://github.com/docker-library/redis/blob/2e14b84ea86939438834a453090966a9bd4367fb/3.0/32bit/docker-entrypoint.sh#L12-L13.

**Comment by timanovsky** [comment]
@yosifkit ok, thanks for explanation !

**Comment by beevelop** [comment]
In case of error logs like these
> Failed opening the RDB file dump.rdb (in server root dir /var/spool/cron) for saving: Permission denied

This seems to be caused by an (attempted) attack to propagate malware: http://blog.huntingmalware.com/notes/LLMalware#attack
Fortunately the script seems to fail at the cronjob injection step.

**Comment by mapleflow** [comment]
hello, in my case.

    # config file not working
    sudo vi /etc/redis/redis.conf
    
    # runtime work in deed
    redis-cli config get dir
    redis-cli config get dbfilename

**Comment by danilosilva87** [comment]
I solved my problem with the @beevelop  friend's tip, as I was just testing, the redis port was open and so I was suffering some attacks

**Comment by bsodmike** [comment]
@danilosilva87 Indeed. For others, here's a quick guide on securing your Docker install http://mwdesilva.com/posts/howto-secure-docker-prevent-leaking-access-to-hackers

This isn't an immediate issue when deploying on AWS/Aliyun due to VPCs inherently being part of the infrastructure; however, on hosts such as DigitalOcean, this does happen.

Interestingly, someone had set a 'password' on one of my redis instances - which, honestly, was a much kinder move by said attacker.

**Comment by danilosilva87** [comment]
Thank you for the link @bsodmike

**Comment by neuronflow** [comment]
I observe the same problem. I am quite confident that no attacker is involved as the machine is in a closed hospital network.
I experimented with owner of the redis folder with everything from the user who spawns the docker container to 999:999. The issue occurs every couple of hours.

**Comment by tianon** [comment]
I would recommend not being so quick to rule out an attacker -- even when employing perimeter defense, sometimes things slip in and then the call is, as they say, "coming from inside the house"

You could try adding a password to your instance and see if the problem persists to help narrow it down.

That being said, these sorts of questions/requests (and any further assistance you require) would be more appropriately posted to [the Docker Community Forums](https://forums.docker.com/), [the Docker Community Slack](https://dockr.ly/slack), or [Stack Overflow](https://stackoverflow.com/search?tab=newest&q=docker).

**Comment by neuronflow** [comment]
Thank you for pointing this out to me. I discovered my redis was actually discoverable even though I thought I deny all connections via ufw and there were some weird requests from an IP near Moscow, Russia:
https://github.com/docker/for-linux/issues/690?fbclid=IwAR2ttgXuWa_qxsT1yuyOd7En74gMpW7A4v0IsaBIuEq83KQ3fW-fL0QVbKg
